File: src/test/java/ru/jerael/booktracker/backend/application/validator/UserValidatorImplTest.java
package ru.jerael.booktracker.backend.application.validator;

import org.junit.jupiter.api.Test;
import ru.jerael.booktracker.backend.domain.exception.ValidationException;
import ru.jerael.booktracker.backend.domain.model.user.UserCreation;
import ru.jerael.booktracker.backend.domain.validator.FieldValidator;
import ru.jerael.booktracker.backend.domain.validator.UserValidator;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

class UserValidatorImplTest {
    private final FieldValidator fieldValidator = new FieldValidatorImpl();
    private final UserValidator userValidator = new UserValidatorImpl(fieldValidator);

    @Test
    void validateCreation_WhenValidDataProvided_ShouldPassValidationWithoutErrors() {
        UserCreation data = new UserCreation("test@example.com", "Password123!");

        assertDoesNotThrow(() -> userValidator.validateCreation(data));
    }

    @Test
    void validateCreation_WhenMultipleInvalidFields_ShouldThrowValidationExceptionWithAllErrors() {
        UserCreation data = new UserCreation("invalid email", "aaa");

        ValidationException exception =
            assertThrows(ValidationException.class, () -> userValidator.validateCreation(data));

        assertThat(exception.getErrors()).hasSizeGreaterThan(1);
        assertThat(exception.getErrors()).anyMatch(e -> e.field().equals("email"));
        assertThat(exception.getErrors()).anyMatch(e -> e.field().equals("password"));
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/application/validator/UserValidatorImplTest.kt
package ru.jerael.booktracker.backend.application.validator

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import ru.jerael.booktracker.backend.domain.exception.ValidationException
import ru.jerael.booktracker.backend.factory.user.UserDomainFactory

class UserValidatorImplTest {
    private val fieldValidator = FieldValidatorImpl()
    private val userValidator = UserValidatorImpl(fieldValidator)
    
    @Test
    fun `when valid data provided, validateCreation should pass validation without errors`() {
        val data = UserDomainFactory.createUserCreation()
        
        assertDoesNotThrow { userValidator.validateCreation(data) }
    }
    
    @Test
    fun `when multiple invalid fields, validateCreation should throw ValidationException with all errors`() {
        val data = UserDomainFactory.createUserCreation(email = "invalid email", password = "aaa")
        
        val exception = assertThrows(ValidationException::class.java) {
            userValidator.validateCreation(data)
        }
        
        with(exception.errors) {
            assertThat(this).hasSizeGreaterThan(1)
            assertThat(this).anyMatch { it.field == "email" }
            assertThat(this).anyMatch { it.field == "password" }
        }
    }
}